src/test/kotlin/me/lensvol/blackconnect/rules.kt

```diff
@@ -0,0 +1,113 @@
+package me.lensvol.blackconnect
+
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.testFramework.PlatformTestUtil
+import com.intellij.testFramework.fixtures.CodeInsightTestFixture
+import com.intellij.testFramework.fixtures.IdeaTestFixtureFactory
+import com.intellij.testFramework.fixtures.impl.LightTempDirTestFixtureImpl
+import com.intellij.util.xmlb.XmlSerializerUtil
+import me.lensvol.blackconnect.settings.BlackConnectProjectSettings
+import org.junit.rules.ExternalResource
+
+/**
+ * Provides `fixture` for code insight and editor tests based on a lightweight reusable IDE project.
+ */
+class IdeaTestRule : ExternalResource() {
+    companion object {
+        init {
+            // Don't show the Dock icon on macOS when running tests
+            System.setProperty("apple.awt.UIElement", "true")
+        }
+    }
+
+    lateinit var fixture: CodeInsightTestFixture
+
+    override fun before() {
+        val factory = IdeaTestFixtureFactory.getFixtureFactory()
+        // We can customize this fixture further here to fit the needs of our plugin tests
+        val fixtureBuilder = factory.createLightFixtureBuilder(null)
+        val tempDirFixture = LightTempDirTestFixtureImpl(true)
+        fixture = factory.createCodeInsightFixture(fixtureBuilder.fixture, tempDirFixture)
+        fixture.testDataPath = FileUtil.toSystemIndependentName(PlatformTestUtil.getCommunityPath())
+        fixture.setUp()
+    }
+
+    override fun after() {
+        fixture.tearDown()
+    }
+}
+
+/**
+ * Runs `blackd` within the project's Poetry environment, waits until it accepts connections before continuing.
+ *
+ * Since blackd takes a couple of seconds to initialise and it is meant to be stateless, you can use this rule
+ * as a `@ClassRule` to save some time.
+ *
+ * This rule installs `blackd` in the Poetry environment of the project if needed.
+ */
+class BlackdProcessRule(private val port: Int, private val timeout: Int = 5_000) : ExternalResource() {
+    private lateinit var process: Process
+
+    override fun before() {
+        installDependencies()
+        val commandLine = GeneralCommandLine("poetry", "run", "blackd", "--bind-port", port.toString())
+        process = commandLine.createProcess()
+        try {
+            BlackdClient("localhost", port, useSsl = false).waitUntilConnected(timeout)
+        } catch (e: Exception) {
+            process.destroyForcibly()
+            throw e
+        }
+    }
+
+    override fun after() {
+        process.destroyForcibly()
+    }
+
+    companion object {
+        private fun installDependencies() {
+            val commandLine = GeneralCommandLine("poetry", "install")
+            val p = commandLine.createProcess()
+            if (p.waitFor() != 0) {
+                error("Error running '${commandLine}': ${p.inputStream.bufferedReader().readText()}")
```

Comment:
If there was an error, it will probably be reported in the output or an error stream. Is `inputStream` really appropriate here?

Reply:
It was my initial thought, but I tested it before committing by running `"poetry", "install", "foo"` and it turns out Poetry reports its error to `stdout`, not `stderr`.

Maybe some of them do go to `stderr` in some cases, so it's safer to show both. I could concatenate `stderr` and `stdout` in my error message. What do you think?

Reply:
No, I mean, wouldn't it make more sense to use _outputStream_ here to get information from **stdout**?

Reply:
Actually `inputStream` is the stdout. Java API is very stupid with names here. From the `Process.getInputStream()` docs:

> Returns the input stream connected to the normal **output** of the process.  The stream obtains data piped from **the standard output** of the process represented by this `Process` object.

Reply:
Ugh. Thank you for clarifying this. :)